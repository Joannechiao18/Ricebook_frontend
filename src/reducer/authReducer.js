// authReducer.js
import { createReducer } from "@reduxjs/toolkit";
import { login, register, logout } from "../actions/authActions";

const initialState = {
  currentUser: null,
  isLoggedIn: false,
  profilePic: null,
};

const authReducer = createReducer(initialState, {
  [login]: (state, action) => {
    state.currentUser = action.payload;
    state.isLoggedIn = true;
    state.profilePic = action.payload.profilePic;
  },
  [register]: (state, action) => {
    state.currentUser = action.payload;
    state.isLoggedIn = true;
    state.profilePic = action.payload.profilePic;
  },
  [logout]: (state) => {
    localStorage.removeItem("persist:root");
    localStorage.removeItem("user");
    localStorage.removeItem("userHeadline");
    state.currentUser = null;
    state.isLoggedIn = false;
  },
});

export default authReducer;

export const selectUser = (state) => state.auth.currentUser;
export const selectProfilePic = (state) => state.auth.profilePic;
