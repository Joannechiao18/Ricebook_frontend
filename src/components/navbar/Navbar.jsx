import "./navbar.scss";
import SearchOutlinedIcon from "@mui/icons-material/SearchOutlined";
import SettingsOutlinedIcon from "@mui/icons-material/SettingsOutlined";
import LogoutOutlinedIcon from "@mui/icons-material/LogoutOutlined";
import { Link, useNavigate } from "react-router-dom";
import { useContext, useState } from "react";
import { Dropdown } from "react-bootstrap";
import { FilterTermContext } from "../../context/FilterTermContext";
import styled from "styled-components";
import { useSelector, useDispatch } from "react-redux";
import { logout } from "../../actions/authActions";
import { selectUser } from "../../reducer/authReducer";

const Navbar = () => {
  const currentUser = useSelector(selectUser);

  const [showDropdown, setShowDropdown] = useState(false);

  const { filterTerm, setFilterTerm } = useContext(FilterTermContext);
  const [searchTerm, setSearchTerm] = useState("");

  const navigate = useNavigate();
  const dispatch = useDispatch();

  const handleInputChange = (e) => {
    setSearchTerm(e.target.value);
  };

  const handleSearchIconClick = () => {
    if (!searchTerm) {
      navigate("/");
    } else {
      setFilterTerm(searchTerm);
    }
  };

  const handleLogout = () => {
    // Your logout logic here

    dispatch(logout());
    navigate("/login");
  };

  return (
    <nav className="navbar navbar-expand-lg navbar-light bg-light sticky-top">
      <div className="container-fluid">
        <div className="d-flex align-items-center">
          <Link className="navbar-brand" to="/">
            Hello World
          </Link>
        </div>

        <div className="d-flex mx-auto">
          <form className="form-inline">
            <input
              className="form-control mr-sm-2"
              type="search"
              placeholder="Search..."
              value={searchTerm}
              onChange={handleInputChange}
            />
            <SearchOutlinedIcon
              onClick={handleSearchIconClick}
              style={{ cursor: "pointer", color: "#938eef" }}
            />
          </form>
        </div>

        <div className="d-flex align-items-center">
          <Dropdown
            className="ms-3 profile-dropdown"
            show={showDropdown}
            onToggle={(isOpen) => setShowDropdown(isOpen)}
          >
            <Dropdown.Toggle
              variant="link"
              id="dropdown-basic"
              className="btn-sm d-flex align-items-center profile-toggle"
            >
              <img
                src={currentUser.avatar}
                alt=""
                className="rounded-cirlce profile-picture"
              />
              <span>{currentUser.username}</span>
            </Dropdown.Toggle>

            <Dropdown.Menu className="custom-dropdown-menu">
              <Dropdown.Item
                className="d-flex justify-content-center align-items-center"
                onClick={() => navigate(`/profile/${currentUser.id}`)}
              >
                <SettingsOutlinedIcon style={{ marginRight: "5px" }} />
                Profile
              </Dropdown.Item>
              <Dropdown.Item
                className="d-flex justify-content-center align-items-center"
                onClick={() => handleLogout()}
              >
                <LogoutOutlinedIcon style={{ marginRight: "5px" }} />
                Logout
              </Dropdown.Item>
            </Dropdown.Menu>
          </Dropdown>
        </div>
      </div>
    </nav>
  );
};

export default Navbar;
