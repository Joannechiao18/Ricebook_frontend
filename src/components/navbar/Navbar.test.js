import React from "react";
import { render, fireEvent } from "@testing-library/react";
import { Provider } from "react-redux";
import { createStore } from "redux";
import Navbar from "../navbar/Navbar";
import authReducer from "../../reducer/authReducer";
import { BrowserRouter as Router } from "react-router-dom";

// Mock the useNavigate hook from react-router-dom
const mockNavigate = jest.fn();
jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockNavigate,
}));

// Mock the useDispatch hook from react-redux
const mockDispatch = jest.fn();
jest.mock("react-redux", () => ({
  ...jest.requireActual("react-redux"),
  useDispatch: () => mockDispatch,
}));

describe("Navbar Component", () => {
  // Before each test, clear all jest mock function calls
  beforeEach(() => {
    mockNavigate.mockClear();
    mockDispatch.mockClear();
  });

  it("logs out the user when the 'Logout' button is clicked", () => {
    // Create a mock Redux store with an initial state
    const mockState = {
      auth: {
        currentUser: {
          id: "1",
          username: "testUser",
          profilePic: "testPic.png",
        },
        isLoggedIn: true,
        profilePic: "testPic.png",
      },
    };

    const store = createStore(authReducer, mockState);

    // Render the Navbar component with the mock store and Router
    const { getByText } = render(
      <Provider store={store}>
        <Router>
          <Navbar />
        </Router>
      </Provider>
    );

    // Simulate user interactions
    const profileDropdown = getByText("testUser");
    fireEvent.click(profileDropdown);

    const logoutButton = getByText("Logout");
    fireEvent.click(logoutButton);

    // Assertions to check expected behavior
    expect(mockDispatch).toHaveBeenCalled();
    expect(mockNavigate).toHaveBeenCalledWith("/login");
  });

  it("matches the snapshot", () => {
    const mockState = {
      auth: {
        currentUser: {
          id: "1",
          username: "testUser",
          profilePic: "testPic.png",
        },
        isLoggedIn: true,
        profilePic: "testPic.png",
      },
    };

    const store = createStore(authReducer, mockState);

    const tree = render(
      <Provider store={store}>
        <Router>
          <Navbar />
        </Router>
      </Provider>
    );

    expect(tree).toMatchSnapshot();
  });
});
