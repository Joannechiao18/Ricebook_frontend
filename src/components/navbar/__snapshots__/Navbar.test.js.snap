// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar Component matches the snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <nav
        class="navbar navbar-expand-lg navbar-light bg-light sticky-top"
      >
        <div
          class="container-fluid"
        >
          <div
            class="d-flex align-items-center"
          >
            <a
              class="navbar-brand"
              href="/"
            >
              Hello World
            </a>
            <div
              class="profile-icon"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="AddReactionOutlinedIcon"
                focusable="false"
                style="font-size: 22px; color: rgb(147, 142, 239);"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 9.5C7 8.67 7.67 8 8.5 8s1.5.67 1.5 1.5S9.33 11 8.5 11 7 10.33 7 9.5zm5 8c2.33 0 4.31-1.46 5.11-3.5H6.89c.8 2.04 2.78 3.5 5.11 3.5zm3.5-6.5c.83 0 1.5-.67 1.5-1.5S16.33 8 15.5 8 14 8.67 14 9.5s.67 1.5 1.5 1.5zM22 1h-2v2h-2v2h2v2h2V5h2V3h-2V1zm-2 11c0 4.42-3.58 8-8 8s-8-3.58-8-8 3.58-8 8-8c1.46 0 2.82.4 4 1.08V2.84C14.77 2.3 13.42 2 11.99 2 6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12c0-1.05-.17-2.05-.47-3H19.4c.38.93.6 1.94.6 3z"
                />
              </svg>
            </div>
          </div>
          <div
            class="d-flex mx-auto"
          >
            <form
              class="form-inline"
            >
              <input
                class="form-control mr-sm-2"
                placeholder="Search..."
                type="search"
                value=""
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="SearchOutlinedIcon"
                focusable="false"
                style="cursor: pointer; color: rgb(147, 142, 239);"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                />
              </svg>
            </form>
          </div>
          <div
            class="d-flex align-items-center"
          >
            <div
              class="ms-3 profile-dropdown dropdown"
            >
              <button
                aria-expanded="false"
                class="btn-sm d-flex align-items-center profile-toggle dropdown-toggle btn btn-link"
                id="dropdown-basic"
                type="button"
              >
                <img
                  alt=""
                  class="rounded-cirlce profile-picture"
                  src="testPic.png"
                />
                <span>
                  testUser
                </span>
              </button>
            </div>
          </div>
        </div>
      </nav>
    </div>
  </body>,
  "container": <div>
    <nav
      class="navbar navbar-expand-lg navbar-light bg-light sticky-top"
    >
      <div
        class="container-fluid"
      >
        <div
          class="d-flex align-items-center"
        >
          <a
            class="navbar-brand"
            href="/"
          >
            Hello World
          </a>
          <div
            class="profile-icon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="AddReactionOutlinedIcon"
              focusable="false"
              style="font-size: 22px; color: rgb(147, 142, 239);"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 9.5C7 8.67 7.67 8 8.5 8s1.5.67 1.5 1.5S9.33 11 8.5 11 7 10.33 7 9.5zm5 8c2.33 0 4.31-1.46 5.11-3.5H6.89c.8 2.04 2.78 3.5 5.11 3.5zm3.5-6.5c.83 0 1.5-.67 1.5-1.5S16.33 8 15.5 8 14 8.67 14 9.5s.67 1.5 1.5 1.5zM22 1h-2v2h-2v2h2v2h2V5h2V3h-2V1zm-2 11c0 4.42-3.58 8-8 8s-8-3.58-8-8 3.58-8 8-8c1.46 0 2.82.4 4 1.08V2.84C14.77 2.3 13.42 2 11.99 2 6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12c0-1.05-.17-2.05-.47-3H19.4c.38.93.6 1.94.6 3z"
              />
            </svg>
          </div>
        </div>
        <div
          class="d-flex mx-auto"
        >
          <form
            class="form-inline"
          >
            <input
              class="form-control mr-sm-2"
              placeholder="Search..."
              type="search"
              value=""
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="SearchOutlinedIcon"
              focusable="false"
              style="cursor: pointer; color: rgb(147, 142, 239);"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
              />
            </svg>
          </form>
        </div>
        <div
          class="d-flex align-items-center"
        >
          <div
            class="ms-3 profile-dropdown dropdown"
          >
            <button
              aria-expanded="false"
              class="btn-sm d-flex align-items-center profile-toggle dropdown-toggle btn btn-link"
              id="dropdown-basic"
              type="button"
            >
              <img
                alt=""
                class="rounded-cirlce profile-picture"
                src="testPic.png"
              />
              <span>
                testUser
              </span>
            </button>
          </div>
        </div>
      </div>
    </nav>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
